### PR TITLE
Add cosign verify-bundle example

### PR DESCRIPTION
**Pull request by danbev**
#### Summary

This commit adds an example that uses `cosign sign-blob` with the `--bundle` option so that a bundle is generated. This bundle will then be used as an argument to verify-bundle, and it will be used to verify a blob.

#### Release Note

NONE

#### Documentation

NONE

Signed-off-by: Daniel Bevenius <daniel.bevenius@gmail.com>

----
I realize that this example is very different from the [verify example](https://github.com/sigstore/sigstore-rs/tree/main/examples/cosign/verify). I've tried to make the usage as simple as possible, but I realize that being new to the code base I might not have enough knowledge about how this could be improved. Hopefully others can provide feedback if there are better ways of doing this.

### CODE REVIEW

File: examples/cosign/verify-bundle/main.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+use tracing_subscriber::{fmt, EnvFilter};
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct Cli {
+    /// Path to bundle file
+    #[clap(short, long)]
+    bundle: String,
+
+    /// Path to artifact to be verify
```
Comment (flavio) [inline examples/cosign/verify-bundle/main.rs:30]:
```suggestion
    /// Path to artifact to be verified
```

File: examples/cosign/verify-bundle/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+This is a simple example that shows how perform cosign verification
```
Comment (flavio) [inline examples/cosign/verify-bundle/README.md:1]: Not a native speaker, hence I might be wrong...

```suggestion
This is a simple example that shows how perform cosign verification
```

File: examples/cosign/verify-bundle/main.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    // setup logging
+    let level_filter = if cli.verbose { "debug" } else { "info" };
+    let filter_layer = EnvFilter::new(level_filter);
+    tracing_subscriber::registry()
+        .with(filter_layer)
+        .with(fmt::layer().with_writer(std::io::stderr))
+        .init();
+
+    let rekor_pub_pem = fs::read_to_string(&cli.rekor_pub_key).unwrap();
```
Comment (flavio) [inline examples/cosign/verify-bundle/main.rs:54]:
```suggestion
    let rekor_pub_pem = fs::read_to_string(&cli.rekor_pub_key).expect("error reading rekor's public key");
```

File: examples/cosign/verify-bundle/main.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    let rekor_pub_key =
+        CosignVerificationKey::from_pem(rekor_pub_pem.as_bytes(), &SigningScheme::default())
+            .expect("Cannot create Rekor verification key");
+    let bundle_json = fs::read_to_string(&cli.bundle).unwrap();
+    let blob = fs::read(&cli.blob.as_str()).unwrap();
```
Comment (flavio) [inline examples/cosign/verify-bundle/main.rs:59]:
```suggestion
    let bundle_json = fs::read_to_string(&cli.bundle).expect("error reading bundle json file");
    let blob = fs::read(&cli.blob.as_str()).expect("error reading blob file");
```

File: examples/cosign/verify-bundle/main.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    let bundle = SignedArtifactBundle::new_verified(&bundle_json, &rekor_pub_key).unwrap();
+    if bundle.verify_blob(&blob).is_ok() {
+        println!("Verification succeeded");
+    } else {
+        println!("Verification failed");
```
Comment (flavio) [inline examples/cosign/verify-bundle/main.rs:65]: nitpicking:

```suggestion
        eprintln!("Verification failed");
        std::process::exit(1);
```

File: src/cosign/bundle.rs [hunk 6]
```diff
@@ -13,9 +13,11 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use base64::{engine::general_purpose, Engine as _};
```
Comment (flavio) [inline src/cosign/bundle.rs:16]: Nitpicking, can you handle the base64 engine like other parts of the project? See https://github.com/sigstore/sigstore-rs/commit/438858dee3bb96d56dd9df5ff0332c62775e9ed4

File: src/cosign/bundle.rs [hunk 7]
```diff
@@ -44,12 +46,24 @@ impl SignedArtifactBundle {
     /// **Note well:** The bundle will be returned only if it can be verified
     /// using the supplied `rekor_pub_key` public key.
     #[allow(dead_code)]
```
Comment (flavio) [inline src/cosign/bundle.rs:48]: This macro can be dropped now that the `new_verified` function is public

**Review by flavio — CHANGES_REQUESTED** [review]
LGTM, I left some small improvements and a change request.

Nice work, thanks for the contribution!

File: examples/cosign/verify-bundle/main.rs (hunk 1 above)
Comment (danbev) [reply examples/cosign/verify-bundle/main.rs]: Updated now, thanks!

File: examples/cosign/verify-bundle/README.md (hunk 2 above)
Comment (danbev) [reply examples/cosign/verify-bundle/README.md]: Updated now, thanks!

File: examples/cosign/verify-bundle/main.rs (hunk 3 above)
Comment (danbev) [reply examples/cosign/verify-bundle/main.rs]: Updated now, thanks!

File: examples/cosign/verify-bundle/main.rs (hunk 4 above)
Comment (danbev) [reply examples/cosign/verify-bundle/main.rs]: Updated now, thanks!

File: src/cosign/bundle.rs (hunk 7 above)
Comment (danbev) [reply src/cosign/bundle.rs]: Updated now, thanks!

File: src/cosign/bundle.rs [hunk 8]
```diff
@@ -44,12 +46,24 @@ impl SignedArtifactBundle {
...
-    pub(crate) fn new_verified(raw: &str, rekor_pub_key: &CosignVerificationKey) -> Result<Self> {
+    pub fn new_verified(raw: &str, rekor_pub_key: &CosignVerificationKey) -> Result<Self> {
         let bundle: SignedArtifactBundle = serde_json::from_str(raw).map_err(|e| {
             SigstoreError::UnexpectedError(format!("Cannot parse bundle |{}|: {:?}", raw, e))
         })?;
         Bundle::verify_bundle(&bundle.rekor_bundle, rekor_pub_key).map(|_| bundle)
     }
+
+    pub fn verify_blob(&self, blob: &[u8]) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (danbev) [reply src/cosign/bundle.rs:57]: Updated now, thanks!

**Review by flavio — APPROVED** [review]
LGTM

File: examples/cosign/verify-bundle/README.md (hunk 2 above)
Comment (flavio) [reply examples/cosign/verify-bundle/README.md]: yes, thanks!

**Comment by flavio** [comment]
Thanks for the contribution and for having addressed the feedback I left during the PR review

**Pull request merged**